src/modules/nats/nats.service.ts

```diff
@@ -1,28 +1,60 @@
-import { Injectable, OnModuleDestroy } from '@nestjs/common';
-import { Client, connect } from 'nats';
+import { Injectable } from '@nestjs/common';
+import { InjectQueue, Process, Processor } from '@nestjs/bull';
+import { Job, Queue } from 'bull';
+import { NatsGateway } from './nats.gateway';
 import { ConfigService } from '@nestjs/config';
 import { Logger } from '../logger/logger.service';
 
+export type IMessageJob = {
+  subject: string;
+  data: Record<string, unknown>;
+};
 @Injectable()
-export class NatsService implements OnModuleDestroy {
-  public connection: Client;
-  constructor(private config: ConfigService, private readonly logger: Logger) {
-    const NATS_CLIENTS_URL = this.config.get<string>('NATS_CLIENTS_URL');
-    try {
-      this.connection = connect(`nats://${NATS_CLIENTS_URL}`, {
-        pingInterval: 5000,
-        maxReconnectAttempts: -1,
-        reconnectDelayHandler: () => {
-          this.logger.log(`Reconnecting to NATS server...`);
-          return 1000;
-        },
-      });
-    } catch (err) {
-      this.logger.error(err, NatsService.name);
-    }
+@Processor('nats-messages')
+export class NatsService {
+  natsSubjectSuffix: string;
+
+  constructor(
+    private natsGateway: NatsGateway,
+    private logger: Logger,
+    private readonly config: ConfigService,
+    @InjectQueue('nats-messages')
+    private readonly messagesQueue: Queue<IMessageJob>,
+  ) {
+    this.natsSubjectSuffix = this.config.get<string>('NATS_SUBJECT_SUFFIX');
+  }
+
+  public async publish(subject: string, data: Record<string, unknown>) {
+    this.logger.debug(
+      `publishing message for subject "${subject}" ${await this.messagesQueue.isReady()}`,
+    );
+    await this.messagesQueue.add('message', {
+      subject,
+      data,
+    });
+    this.logger.debug(`published message for subject "${subject}"`);
+  }
+
+  public async publishForDids(
+    requestType: string,
+    topic: string,
+    dids: string[],
+    data: Record<string, unknown>,
+  ) {
+    await Promise.all(
+      dids.map(did =>
+        this.publish(
+          `${requestType}.${topic}.${did}.${this.natsSubjectSuffix}`,
+          data,
+        ),
+      ),
+    );
   }
 
-  async onModuleDestroy() {
-    this.connection.close();
+  @Process('message')
+  public async processMessage(job: Job<IMessageJob>) {
+    const { subject, data } = job.data;
+    this.natsGateway.connection.publish(subject, JSON.stringify(data));
+    this.natsGateway.connection.flush();
```

Comment:
Is this a best practice to `flush` after a `publish`?

Reply:
Jakub and Kris investigated the problem. It seems that without flush there is a risk of not sending the message at all.

Reply:
@Harasz @hejkerooo when you have a moment, could you comment with any additional information you have about why a `flush` after `publish` is necessary?

Reply:
https://docs.nats.io/using-nats/developer/sending/caches

> For performance reasons, most if not all, of the client libraries will buffer outgoing data so that bigger chunks can be written to the network at one time. This may be as simple as a byte buffer that stores a few messages before being pushed to the network.

@jrhender

Reply:
@Harasz Hmm, interesting, thanks. You'd think NATS would send the data as a small chunk after a certain amount if it wasn't able to assemble a bigger chunk. So this is kind-of a workaround past a NATS performance optimization 🤔 ?

Reply:
Using flush after publish in this situation we are publishing event immediately. We will not wait until the buffer is full. This results in disabling default NATS optimization, but also event about claims will be published immediately.